pkg/storage/local.go

```diff
@@ -0,0 +1,482 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	"io"
+	llog "log"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+
+	"github.com/TierMobility/boring-registry/pkg/core"
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+)
+
+type LocalFileSystem interface {
+	OpenFile(name string, flag int, perm os.FileMode) (io.ReadWriteCloser, error)
+	ReadFile(name string) ([]byte, error)
+	Stat(name string) (os.FileInfo, error)
+	ReadDir(name string) ([]os.DirEntry, error)
+	MkdirAll(name string, perm os.FileMode) error
+}
+
+type FileServer interface {
+	ListenAndServe() error
+	Addr() string
+}
+
+type LocalStorage struct {
+	fs                  LocalFileSystem
+	server              FileServer
+	serverEndpoint      string
+	storageDir          string
+	moduleArchiveFormat string
+}
+
+func NewLocalStorage(fs LocalFileSystem, server FileServer, storageDir, serverEndpoint, moduleArchiveFormat string) *LocalStorage {
+	// when use client to upload or migrate, no need to set up the http file server
+	if server != nil && !reflect.ValueOf(server).IsNil() {
+		go func() {
+			if err := server.ListenAndServe(); err != nil {
+				llog.Printf("error: %+v", err)
```

Comment:
Thanks for your responsible advise and i get your point. But I still have some doubts about the examples you mentioned earlier.
https://github.com/TierMobility/boring-registry/blob/62960773833a865f8f7b4b02414a1281c512b1f2/cmd/server.go#L113-L122
when cancel function is called, channel returned by ctx.Done() has been closed, but it is still passed to the shutdown function. In this situation, will shutdown return immediately and not wait to close the connection actually?

Reply:
That's a good catch. thanks! As you pointed out, the `shutdown` does not wait to close the connections in this case.
I think it would be better to create a new context with e.g. a deadline of 10s:
```go
ctx, cancelCtx := context.WithDeadline(context.Background(), 10*time.Second)
defer cancelCtx()
```